### DISCUSSION TITLE
Backup/Restore

### DISCUSSION THREAD

**Discussion by jmatsushita** [kubevela/kubevela#1645] [Q&A]
Hello,

I'm wondering whether backup and restore is part of the OAM model and something that is considered in scope for KubeVela? Or is the idea that something like Velero would actually be able to provide backup/restore capabilities? Would it then need to also be configured to backup the KubeVela CRDs? Examples of how this could be achieved would be really useful!

Cheers,

Jun

**Comment by wonderflow** [comment]
Sorry for the late reply! I think part of the `backup/restore` thing is in OAM model.

1. Configure a trait to periodically backup the resources of the component.
2. Configure a scope to periodically backup all the resources of applications included.

While the trait or scope still rely on the project like Velero to provide capability.

**Comment by hprotzek** [comment]
Hello, we are working on a backup trait with Velero, to backup a complete deployed kubevela app.
We are using the selector to select all resources belonging to an component:

```
velero backup create backup_1 --selector app.oam.dev/name=my-vela-app --include-cluster-resources=false
```

The actual backup and restore works and the app including volumes are working after the restore, but `vela ls` doesn't see it. 
Which vela oam resources do we need to backup to get this working?

**Reply by Somefive** [reply]
Good to hear that velero could do the backup and restore! 

Except for `vela ls`, does other commands work? Like `vela status` or `vela top`? Have your applications entered the running state?

By using `--include-cluster-resources=false`, does it mean that cluster-scoped resource with the label `app.oam.dev/name=my-vela-app` will not be recovered? We have a underlying resource named ResourceTracker which is a cluster-scoped resource and it is necessary for it to be migrated for applications to work. (It is cluster-scoped because of the history compatibility issues.)

**Reply by RemkovZ** [reply]
It looks like Kubevela adds a `app.oam.dev/name` label to the resources it creates, except for the `Application` resource. The `Application` resource only gets a `oam.dev/render-hash` label.

Label on `Application`:
```
$ k get app my-vela-app -o json | jq .metadata.labels
{
  "oam.dev/render-hash": "f7de830122e1d922"
}
```

Labels on `Deployment`:
```
$ k get deploy my-vela-app -o json | jq .metadata.labels
{
...
  "app.oam.dev/name": "my-vela-app",
...
}
```

Since Velero is backing up resources using a label selector, it is missing the `Application` resource. If we manually add the label `app.oam.dev/name=my-vela-app` on the `Application` resource, let Velero create a backup and restore that backup, then `vela ls` is showing the app.

Is there a way we can have Kubevela add the `app.oam.dev/name` also on the `Application` resource?

**Reply by wonderflow** [reply]
You can just add `app.oam.dev/name` label to the application when you creating it. The `Application` resource is created by user not generated.

**Reply by RemkovZ** [reply]
Yes, that is clear, but that is error prone, since user can make a typo in the label and then the backup is still not working correctly.

Kubevela seems to be setting the `render-hash` label on the `Application` resource, so it is touching the resource in some way, or is it not?

**Reply by wonderflow** [reply]
No, not all applications have the `render-hash` label, it only exists in some cases which can be generate by velaux/vela-cli. Application YAML should be controlled by users in general case. 

But you're right, it can provide convenience when it can be generated automatically, we can add the `app.oam.dev/name` in mutating webhook.  @Somefive  WDYT?

**Reply by Somefive** [reply]
Usually adding label `app.oam.dev/name` to application works but it can encounter issues when there are some applications managing other applications. So the `app.oam.dev/name` label on application actually means the owner app of that application. So I would not suggest using this label directly on applications, but in usual cases it is fine for you to do that manually.

`render-hash` is not always attached to resources so it is suggested not to rely on it.

I would recommend more to backup the application aside by its resources. BTW, `--include-cluster-resources=false` does it means cluster scoped resource will not be included in the backup. If so, there could be something wrong since one of the required system resource `ResourceTracker` is cluster-scoped.

**Reply by RemkovZ** [reply]
Thanks for all the feedback so far!

Instead of trying to backup the application and all related resources, we decided to focus on backups of the persistent volume and provide a full self service (backup/restore) around that and take it from there.

**Reply by wonderflow** [reply]
@RemkovZ We'd like to hear your solution if you can share it to the community. The backup/restore feature can be helpful to the whole community.

**Reply by charlie0129** [reply]
Not being able to add labels to the Application means Velero cannot select the Application itself for backups. Some workarounds may be needed. I think this is mainly blocked by Velero not being able to select individual resources. According what the Velero team says, similar features should already be planned https://github.com/vmware-tanzu/velero/issues/5652 . By the time when Velero added the ability to select individual resources to backup, I think this issue should be easier to solve.
